serialization.rst

```diff
@@ -0,0 +1,308 @@
+to inject your own ``__pclass`` array element.
+
+Examples
+~~~~~~~~
+
+::
+
+  stdClass {
+    public $foo => 42
+  } => { 'foo' : 42 }
+
+  MyClass {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+  } => { 'foo' : 42 }
+
+  AnotherClass implements MongoDB\BSON\Serializable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine" }
+
+  UpperClass implements MongoDB\BSON\Persistable {
+    public $foo => 42,
+    protected $prot => "wine",
+    private $fpr => "cheese"
+    function bsonSerialize() : array {
+        return [ 'foo' => $this->foo, 'prot' => $this->prot ];
+    }
+  } => { 'foo' : 42, 'prot' : "wine", '__pclass' : MongoDB\BSON\Binary(0x80, "UpperClass") }
+
+
+Deserialization from BSON
+=========================
+
+For compound types, there are three data types:
+
+- ``root``: refers to the top-level BSON document *only*
+- ``document``: refers to embedded BSON documents *only*
+- ``array``: refers to a BSON array
+
+Each of those three data types can be mapped against different PHP types. The
+possible mapping values are:
+
+- *not set* or ``NULL`` — this is the default.
+
+  - A BSON array will be deserialized as a PHP ``array``.
+  - A BSON document (root or embedded) without ``__pclass`` property [1]_ becomes a
+    PHP ``stdClass`` object, with each BSON document key becoming a
+    public ``stdClass`` property.
+  - A BSON document (root or embedded) with ``__pclass`` property [1]_ becomes
+    a PHP object of the class name as defined by the ``__pclass`` property.
+
+    If the named class implements the ``MongoDB\BSON\Unserializable``
+    interface, then the properties of the BSON document, including the
+    ``__pclass`` property, are sent as an associative array to the
+    ``bsonUnserialize`` function to initialise the object's properties.
+    
+    If the named class does not implement the ``MongoDB\BSON\Unserializable``
+    interface, then an ``MongoDB\Driver\UnexpectedValueException`` exception
+    is thrown with a message indicating that the class does not implement the
+    expected interface.
+
+- ``"array"`` — turns a BSON array or BSON document into a PHP array.
+  ``__pclass`` properties [1]_ are ignored, and are **not** set as array
+  element in the returned array.
+
+- ``"object"`` or ``"stdClass"`` — turns a BSON array or BSON document into a
+  ``stdClass`` object. There will be no special treatment of a ``__pclass``
+  property [1]_, but it should **not** be set as property in the returned
+  object.
+
+- ``any other string`` — defines the class name that the BSON array or BSON
+  object should be deserialized at.
+
+  If the class implements the ``MongoDB\BSON\Unserializable`` interface,
+  either directly, or indirectly via ``MongoDB\BSON\Persistable``, then
```

Comment:
[Google tells me](https://www.google.com/webhp?q=%22either%20directly%20or%20indirectly%20via%22) we don't need a comma after "directly".